apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,378 @@
+}
+
+// StageConfig describes a single update stage group configuration.
+// The clusters in each stage are updated sequentially for now.
+// We will stop the update if any of the updates fail.
+type StageConfig struct {
+	// The name of the stage. This MUST be unique within the same StagedUpdateStrategy.
+	// +required
+	Name string `json:"name"`
+
+	// LabelSelector is a label query over all the joined member clusters. Clusters matching the query are selected.
+	// We don't check if there are overlap between clusters selected by different stages.
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
+
+	// The label key used to sort the selected clusters.
+	// The clusters within the stage are updated sequentially following the rule below:
+	//   - primary: Ascending order based on the value of the label key, interpreted as integers if present.
+	//   - secondary: Ascending order based on the name of the cluster if the label key is absent.
+	// +optional
+	SortingLabelKey *string `json:"sortingLabelKey,omitempty"`
+
+	// The collection of tasks that each stage needs to complete successfully before moving to the next stage.
+	// Each task is executed in parallel and they can not be more than one task of the same type.
+	// +optional
+	AfterStageTasks []AfterStageTask `json:"afterStageTasks,omitempty"`
+}
+
+// AfterStageTask is the collection of post stage tasks that ALL need to be completed before we can move to the next stage.
+type AfterStageTask struct {
+	// The type of the after stage task.
+	// +kubebuilder:validation:Enum=TimedWait;Approval
+	// +required
+	Type AfterStageTaskType `json:"type"`
+
+	// The prefix of the name of the StageApprovalRequest we need to create to get the approval for the next stage.
+	// We create the full name of the approval request by concatenating the prefix to the stagedUpdateRunName and the stageName.
+	// +optional
+	ApprovalRequestPrefix string `json:"approvalTask,omitempty"`
+
+	// The time to wait after all the clusters in the current stage complete the update before we move to the next stage.
+	// +kubebuilder:default:="1h"
+	// +kubebuilder:validation:Pattern="^0|([0-9]+(\\.[0-9]+)?(s|m|h))+$"
+	// +kubebuilder:validation:Type:=string
+	// +optional
+	WaitTime metav1.Duration `json:"waitTime,omitempty"`
+}
+
+// StagedUpdateRunStatus defines the observed state of the StagedUpdateRun.
+type StagedUpdateRunStatus struct {
+	// UpdateRunStrategy is the collection of the stageConfig in the StagedUpdateStrategy that we are going to use for the update.
+	// +required
+	UpdateRunStrategy []StageConfig `json:"updateRunStrategy"`
+
+	// AppliedStrategy is applied strategy of the CRP that we are currently using.
+	// It is the same as the strategy in the CRP when we first start the update.
+	// We will NOT update the strategy in the update even if the appliedStrategy changes in the CRP.
+	// +required
+	AppliedStrategy v1beta1.ApplyStrategy `json:"appliedStrategy"`
```

Comment:
ApplyStrategy instead of AppliedSrategy? AppliedStrategy is too generic. L159 is also an applied strategy.